src/main/java/bio/terra/cli/Resource.java

```diff
@@ -0,0 +1,229 @@
+   * Calls the appropriate sub-class constructor based on the resource type.
+   */
+  public static Resource deserializeFromWsm(ResourceDescription wsmObject) {
+    bio.terra.workspace.model.ResourceType wsmResourceType =
+        wsmObject.getMetadata().getResourceType();
+    switch (wsmResourceType) {
+      case GCS_BUCKET:
+        return new GcsBucket(wsmObject);
+      case BIG_QUERY_DATASET:
+        return new BqDataset(wsmObject);
+      case AI_NOTEBOOK:
+        return new AiNotebook(wsmObject);
+      default:
+        throw new IllegalArgumentException("Unexpected resource type: " + wsmResourceType);
+    }
+  }
+
+  /**
+   * Serialize the internal representation of the resource to the format for command input/output.
+   */
+  public abstract CommandResource serializeToCommand();
+
+  /** Serialize the internal representation of the resource to the format for writing to disk. */
+  public abstract DiskResource serializeToDisk();
+
+  /**
+   * Check if the name only contains alphanumeric and underscore characters. Sync the cached list of
+   * resources.
+   *
+   * <p>When launching an app, either in a Docker container or a local process, we pass a map of all
+   * the resources in the workspace to their resolved cloud identifiers (e.g. TERRA_MYBUCKET ->
+   * gs://mybucket). This is the reason for this restriction at resource creation time.
+   *
+   * @param name string to check
+   * @return true if the string is a valid environment variable name
+   */
+  protected static boolean isValidEnvironmentVariableName(String name) {
+    return !Pattern.compile("[^a-zA-Z0-9_]").matcher(name).find();
+  }
+
+  /** Delete an existing resource in the workspace. */
+  public void delete() {
+    switch (stewardshipType) {
+      case REFERENCED:
+        deleteReferenced();
+        break;
+      case CONTROLLED:
+        deleteControlled();
+        break;
+      default:
+        throw new IllegalArgumentException("Unknown stewardship type: " + stewardshipType);
+    }
+    listAndSync();
+  }
+
+  /** Call WSM to delete a referenced resource. */
+  protected abstract void deleteReferenced();
+
+  /** Call WSM to delete a controlled resource. */
+  protected abstract void deleteControlled();
+
+  /** Resolve a resource to its cloud identifier. */
+  public abstract String resolve();
+
+  /**
+   * Helper enum for the {@link #checkAccess(CheckAccessCredentials)} method. Specifies whether to
+   * use end-user or pet SA credentials for checking access to a resource in the workspace.
+   */
+  public enum CheckAccessCredentials {
+    USER,
+    PET_SA;
+  };
+
+  /** Call WSM to check whether the current user has access to a resource. */
+  public abstract boolean checkAccess(CheckAccessCredentials credentialsToUse);
+
+  /** Fetch the list of resources for the current workspace. Sync the cached list of resources. */
+  public static List<Resource> listAndSync() {
```

Comment:
Hmm, I put listing resources in the `Resource` class because listing workspaces is in the `Workspace` class. But I think you're right it makes more sense to put both in the `Workspace` class.